flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java

```diff
@@ -0,0 +1,353 @@
+
+            try {
+                for (int i = 0; i < numTotalBuffers; ++i) {
+                    buffers.add(MemorySegmentFactory.allocateUnpooledOffHeapMemory(bufferSize));
+                }
+            } catch (OutOfMemoryError outOfMemoryError) {
+                int allocated = buffers.size();
+                buffers.clear();
+                throw new OutOfMemoryError(
+                        String.format(
+                                "Can't allocate enough direct buffer for batch shuffle read buffer "
+                                        + "pool (bytes allocated: %d, bytes still needed: %d). To "
+                                        + "avoid the exception, you need to do one of the following"
+                                        + " adjustments: 1) If you have ever decreased %s, you need"
+                                        + " to undo the decrement; 2) If you ever increased %s, you"
+                                        + " should also increase %s; 3) If neither the above cases,"
+                                        + " it usually means some other parts of your application "
+                                        + "have consumed too many direct memory and the value of %s"
+                                        + " should be increased.",
+                                allocated * bufferSize,
+                                (numTotalBuffers - allocated) * bufferSize,
+                                TaskManagerOptions.FRAMEWORK_OFF_HEAP_MEMORY.key(),
+                                TaskManagerOptions.NETWORK_BATCH_SHUFFLE_READ_MEMORY.key(),
+                                TaskManagerOptions.FRAMEWORK_OFF_HEAP_MEMORY.key(),
+                                TaskManagerOptions.TASK_OFF_HEAP_MEMORY.key()));
+            }
+        }
+    }
+
+    /**
+     * Requests a collection of buffers (determined by {@link #numBuffersPerRequest}) from this
+     * buffer pool. Exception will be thrown if no enough buffers can be allocated in the given
+     * timeout and the corresponding requester is not holding any allocated buffers currently.
+     */
+    public List<MemorySegment> requestBuffers(Object owner) throws Exception {
+        checkArgument(owner != null, "Owner must be not null.");
+
+        List<MemorySegment> allocated = new ArrayList<>(numBuffersPerRequest);
+        synchronized (buffers) {
+            checkState(!destroyed, "Buffer pool is already destroyed.");
+
+            if (!initialized) {
+                initialize();
+            }
+
+            long startTime = System.currentTimeMillis();
+            Counter counter = numBuffersAllocated.get(owner);
+            while (buffers.size() < numBuffersPerRequest
+                    || (counter != null
+                            && counter.get() + numBuffersPerRequest > maxBuffersPerRequester)) {
+                checkState(!destroyed, "Buffer pool is already destroyed.");
+
+                buffers.wait(requestTimeout);
```

Comment:
seems `buffers` is not notified when `numBuffersAllocated` is updated.

Reply:
Do you mean when the all buffers are recycled and the allocated buffers reaches 0. If that is the case, the timeout will be triggered in the next round. Currently, we only notify when there is enough buffers or the buffer pool is destroyed.

Reply:
I'm afraid we might still need the notification since it is also waiting for the condition `(counter != null
&& counter.get() + numBuffersPerRequest > maxBuffersPerRequester)`, if there is no notification for this condition, after there are enough buffers, the thread might still blocked on this condition.

Reply:
The recycle thread will always notify the lock

Reply:
close this comment because it is outdated with the removal of `maxBuffersPerRequester`